cpp/daal/src/algorithms/spectral_embedding/spectral_embedding_default_dense_impl.i

```diff
@@ -0,0 +1,214 @@
+#include "src/externals/service_blas.h"
+#include "src/threading/threading.h"
+#include "src/algorithms/service_error_handling.h"
+#include "src/data_management/service_numeric_table.h"
+#include "src/algorithms/cosdistance/cosdistance_kernel.h"
+#include "src/externals/service_lapack.h"
+#include <iostream>
+
+using namespace daal::internal;
+
+namespace daal
+{
+namespace algorithms
+{
+namespace spectral_embedding
+{
+namespace internal
+{
+
+template <typename algorithmFPType, CpuType cpu>
+services::Status computeEigenvectorsInplace(size_t nFeatures, algorithmFPType * eigenvectors, algorithmFPType * eigenvalues)
+{
+    char jobz = 'V';
+    char uplo = 'U';
+
+    DAAL_INT lwork  = 2 * nFeatures * nFeatures + 6 * nFeatures + 1;
+    DAAL_INT liwork = 5 * nFeatures + 3;
+    DAAL_INT info;
+
+    TArray<algorithmFPType, cpu> work(lwork);
+    TArray<DAAL_INT, cpu> iwork(liwork);
+    DAAL_CHECK_MALLOC(work.get() && iwork.get());
+
+    LapackInst<algorithmFPType, cpu>::xsyevd(&jobz, &uplo, (DAAL_INT *)(&nFeatures), eigenvectors, (DAAL_INT *)(&nFeatures), eigenvalues, work.get(),
+                                             &lwork, iwork.get(), &liwork, &info);
+    if (info != 0) return services::Status(services::ErrorPCAFailedToComputeCorrelationEigenvalues); // CHANGE ERROR STATUS
+    return services::Status();
+}
+
+/**
+ *  \brief Kernel for Spectral Embedding calculation
+ */
+template <typename algorithmFPType, Method method, CpuType cpu>
+services::Status SpectralEmbeddingKernel<algorithmFPType, method, cpu>::compute(const NumericTable * xTable, NumericTable * embeddingTable,
+                                                                                NumericTable * eigenTable, const KernelParameter & par)
+{
+    services::Status status;
+    // std::cout << "inside DAAL kernel" << std::endl;
+    // std::cout << "Params: " << par.numberOfEmbeddings << " " << par.numberOfNeighbors << std::endl;
+    size_t k       = par.numberOfEmbeddings;
+    size_t filtNum = par.numberOfNeighbors + 1;
+    size_t n       = xTable->getNumberOfRows(); /* Number of input feature vectors   */
+
+    SharedPtr<HomogenNumericTable<algorithmFPType> > tmpMatrixPtr =
+        HomogenNumericTable<algorithmFPType>::create(n, n, NumericTable::doAllocate, &status);
+
+    DAAL_CHECK_STATUS_VAR(status);
+    NumericTable * covOutput   = tmpMatrixPtr.get();
+    NumericTable * a0          = const_cast<NumericTable *>(xTable);
+    NumericTable * eigenvalues = const_cast<NumericTable *>(eigenTable);
+
+    // Compute cosine distances matrix
+    {
+        auto cosDistanceKernel = cosine_distance::internal::DistanceKernel<algorithmFPType, cosine_distance::Method::defaultDense, cpu>();
+        DAAL_CHECK_STATUS(status, cosDistanceKernel.compute(0, &a0, 0, &covOutput, nullptr));
+    }
+
+    WriteRows<algorithmFPType, cpu> xMatrix(covOutput, 0, n);
+    DAAL_CHECK_BLOCK_STATUS(xMatrix);
+    algorithmFPType * x = xMatrix.get();
+
```

Comment:
I believe that according to SDL somewhere around here should be several `DAAL_ASSERT_...` statements to check if your access to the memory is contained.